contracts/src/protocol/XRegistry.sol

```diff
@@ -0,0 +1,191 @@
+import { XRegistryReplica } from "./XRegistryReplica.sol";
+
+/**
+ * @title XRegistry
+ * @notice Registry for cross-chain contract deployments. Replicated across each supported chain.
+ * @dev Using Ownable, rather than OwnableUpgradeable, to because predeploys don't require initalization.
+ */
+contract XRegistry is Ownable, XRegistryBase {
+    /**
+     * @notice Register a contract deployment.
+     */
+    event ContractRegistered(uint64 indexed chainId, string indexed name, address indexed registrant, address addr);
+
+    /**
+     * @notice xcall gas limit for XRegistryReplica.set
+     */
+    uint64 public constant XSET_GAS_LIMIT = 100_000;
+
+    /**
+     * @notice OmniPortal contract.
+     */
+    OmniPortal public omni;
+
+    /**
+     * @notice Mapping of chain IDs to the XRegistryReplica address on that chain.
+     */
+    mapping(uint64 => address) public replicas;
+
+    /**
+     * @notice List of supported chain IDs (ie. chain ids with XRegistry replicas)
+     */
+    uint64[] internal _chainIds;
+
+    /**
+     * @notice Register a contract deployment.
+     */
+    function register(uint64 chainId, string calldata name, address addr) external payable {
+        require(isSupportedChain(chainId), "XRegistry: chain not supported");
+        require(msg.value >= _syncFee(chainId, name, msg.sender, addr), "XRegistry: insufficient fee");
+
+        _set(chainId, name, msg.sender, addr);
+        _sync(chainId, name, msg.sender, addr);
+
+        emit ContractRegistered(chainId, name, msg.sender, addr);
+    }
+
+    /**
+     * @notice Calculate the fee to register a contract.
+     */
+    function registrationFee(uint64 chainId, string calldata name, address addr) external view returns (uint256) {
+        return _syncFee(chainId, name, msg.sender, addr);
+    }
+
+    /**
+     * @notice Syncs replicas with a new registration for contract `name` by `registrant`.
+     *          - Adds all existing registrations of `name` by `registrant` to the replica on `chainId`
+     *          - Adds the new deployment on `chainId` to all replicas with existing registrations
+     *            of `name` by `registrant`.
+     *
+     *          Ex. Consider contract with name "MyContract" and registrant "0x1234". It has 2
+     *          deployments.
+     *              - Chain 1, address 0x1111
+     *              - Chain 2, address 0x2222
+     *
+     *          When registering a new deployment on Chain 3, address 0x3333, the sync process is as
+     *          follows.
+     *              - Add (Chain 1 -> 0x1111) and (Chain 2 -> 0x2222) to Chain 3
+     *              - Add (Chain 3 -> 0x3333) to Chain 1 and Chain 2
+     */
+    function _sync(uint64 chainId, string calldata name, address registrant, address addr) internal {
+        for (uint256 i = 0; i < _chainIds.length; i++) {
+            uint64 otherChainId = _chainIds[i];
+
+            // don't sync with self
+            if (otherChainId == chainId) continue;
```

Comment:
don't we need to sync with self if there's a replica on this chain?

Reply:
Sorry think I mis understood - there is not "replica" on omni EVM. the `XRegistry` itself acts as the replica to the omni evm portal.